databricks/sdk/credentials_provider.py

```diff
@@ -664,6 +665,86 @@ def inner() -> Dict[str, str]:
+        """
+        val = (
+            os.environ.get("IS_IN_DB_MODEL_SERVING_ENV")
+            # Checking the old env var name for backward compatibility. The env var was renamed once
+            # to fix a model loading issue, but we still need to support it for a while.
+            or os.environ.get("IS_IN_DATABRICKS_MODEL_SERVING_ENV") or "false")
+        return val.lower() == "true"
+
+    @classmethod
+    def should_fetch_model_serving_environment_oauth(cls) -> bool:
+        """
+        Check whether this is the model serving environment
+        Additionally check if the oauth token file path exists
+        """
+        return (cls.is_in_databricks_model_serving_environment()
+                and os.path.isfile(cls._MODEL_DEPENDENCY_OAUTH_TOKEN_FILE_PATH))
+
+    @classmethod
+    def get_model_dependency_oauth_token(cls, should_retry=True) -> str:
+        try:
+            with open(cls._MODEL_DEPENDENCY_OAUTH_TOKEN_FILE_PATH) as f:
+                oauth_dict = json.load(f)
+                return oauth_dict["OAUTH_TOKEN"][0]["oauthTokenValue"]
+        except Exception as e:
+            # sleep and retry in case of any race conditions with OAuth refreshing
+            if should_retry:
+                time.sleep(0.5)
+                return cls.get_model_dependency_oauth_token(should_retry=False)
+            else:
+                raise RuntimeError("Unable to read Oauth credentials from file mount for Databricks "
+                                   "Model Serving dependency failed") from e
+
+    @classmethod
+    def get_databricks_host_token(cls) -> Optional[Tuple[str, str]]:
+        if not cls.should_fetch_model_serving_environment_oauth():
+            return None
+        MODEL_SERVING_HOST_ENV_VAR = "DATABRICKS_MODEL_SERVING_HOST_URL"
+        DB_MODEL_SERVING_HOST_ENV_VAR = "DB_MODEL_SERVING_HOST_URL"
+
+        # read from DB_MODEL_SERVING_HOST_ENV_VAR if available otherwise MODEL_SERVING_HOST_ENV_VAR
+        host = os.environ.get(DB_MODEL_SERVING_HOST_ENV_VAR) or os.environ.get(MODEL_SERVING_HOST_ENV_VAR)
+        token = cls.get_model_dependency_oauth_token()
+
+        return (host, token)
+
+
+@credentials_strategy('model-serving', [])
+def model_serving_auth(cfg: 'Config') -> Optional[CredentialsProvider]:
+    try:
+        host, token = ModelServingAuthProvider.get_databricks_host_token()
+        if token is None:
+            raise ValueError("Unable to Authenticate using Model Serving Environment")
+        if cfg.host is None:
+            cfg.host = host
+    except Exception as e:
+        logger.info("Unable to use Databricks Model Serving Environment " + str(e))
+        return None
+
+    logger.info("Using Databricks Model Serving Authentication")
+
+    def inner() -> Dict[str, str]:
+        # Call here again to get the refreshed token
+        host, token = ModelServingAuthProvider.get_databricks_host_token()
```

Comment:
This method is called with every request. Do we need to check the file every time?  We incur the cost of opening a file & reading its contents on every request. Is there an expiry included in the token file? We could use that to determine whether we need to refresh the token.

I tested a simple benchmark:

```py
import json
import timeit
import os

with open('sample.json', 'w') as f:
    json.dump([{'key': 'value'} for _ in range(100)], f)

def benchmark():
    with open('sample.json', 'r') as f:
        data = json.load(f)

print(f"Average time to open, read, and parse file: {timeit.timeit(benchmark, number=1000):.6f} seconds")
```
This came out to 0.07 seconds on my laptop. It might be nice to avoid this overhead, especially for latency-sensitive use-cases.

Reply:
Good call, added a 5 min cache